src/app/code/community/Smartdevs/Indexer/Model/Resource/Indexer/Stock.php

```diff
@@ -0,0 +1,154 @@
+<?php
+
+/**
+ * CatalogInventory Stock Status Indexer Resource Model
+ *
+ * @category    Smartdevs
+ * @package     Smartdevs_Indexer
+ * @author      Daniel Niedergesäß <daniel.niedergesaess@gmail.com>
+ */
+class Smartdevs_Indexer_Model_Resource_Indexer_Stock extends Mage_CatalogInventory_Model_Resource_Indexer_Stock
+{
+    /**
+     * retrieve helper instance
+     *
+     * @return Smartdevs_Indexer_Helper_Data
+     */
+    protected function getIndexerHelper()
+    {
+        return Mage::helper('smartdevs_indexer');
+    }
+
+    /**
+     * Retrieve new index table name
+     *
+     * @return string
+     */
+    private function getNewTable()
+    {
+        return $this->getMainTable() . '_new';
+    }
+
+    /**
+     * Retrieve old index table name
+     *
+     * @return string
+     */
+    private function getOldTable()
+    {
+        return $this->getMainTable() . '_old';
+    }
+
+    /**
+     * Clean up temporary index table
+     *
+     * magento runs per default delete from table which blows up
+     * the mysql transaction log
+     * this method isn't required anymore
+     */
+    public function clearTemporaryIndexTable()
+    {
+        return $this;
+    }
+
+    /**
+     * Synchronize data between index storage and original storage
+     *
+     * we use here a table rotation instead of deleting the whole table inside an transaction
+     * which blows up the mysql transaction log and creates a lot of iops inside the database
+     *
+     * @return Mage_Index_Model_Resource_Abstract
+     */
+    public function syncData()
+    {
+        //clean up last rotation if there was an error
+        $this->_getWriteAdapter()->dropTable($this->getNewTable());
+        $this->_getWriteAdapter()->dropTable($this->getOldTable());
+
+        try {
+            $table = $this->_getWriteAdapter()->createTableByDdl($this->getMainTable(), $this->getNewTable());
+
+            //foreign keys should be unique so we need to change the names of the table
```

Comment:
suggestion: extract this block into a renameForeignKeys() method (lower level of abstraction, separate impl. details from logic)